external/fv3fit/fv3fit/keras/_models/precipitative.py

```diff
@@ -0,0 +1,465 @@
+            activation="linear",
+            activity_regularizer=self._residual_regularizer,
+            name="dense_output",
+        )(norm_x)
+        denormalized_output = self.output_without_precip_scaler.denormalize_layer(
+            norm_output_vector
+        )
+        unpacked_output = get_unpack_layer(
+            self.output_without_precip_packer, feature_dim=1
+        )(denormalized_output)
+        assert self.output_without_precip_packer.pack_names[0] == "dQ1"
+        T_tendency = unpacked_output[0]
+        q_tendency = unpacked_output[1]
+
+        norm_column_precip_vector = tf.keras.layers.Dense(
+            self.humidity_packer.feature_counts[self._Q_TENDENCY_NAME],
+            activation="linear",
+        )(norm_x)
+        column_precip = self.humidity_scaler.denormalize_layer(
+            norm_column_precip_vector
+        )
+        column_heating = tf.keras.layers.Lambda(condensational_heating)(column_precip)
+        if not self._dense_behavior:
+            T_tendency = tf.keras.layers.Add(name="T_tendency")(
+                [T_tendency, column_heating]
+            )
+            q_tendency = tf.keras.layers.Add(name="q_tendency")(
+                [q_tendency, column_precip]
+            )
+        surface_precip = tf.keras.layers.Add(name="add_physics_precip")(
+            [
+                physics_precip,
+                tf.keras.layers.Lambda(integrate_precip, name="surface_precip")(
+                    [column_precip, delp]
+                ),
+            ]
+        )
+        assert list(self.output_variables) == [
+            "dQ1",
+            "dQ2",
+            "total_precipitation_rate",
+        ], self.output_variables
+        train_model = tf.keras.Model(
+            inputs=input_layers,
+            outputs=(T_tendency, q_tendency, surface_precip)
+            + tuple(unpacked_output[2:]),
+        )
+        train_model.compile(
+            optimizer=self._optimizer,
+            loss=get_losses(
+                self.output_variables, self.output_packer, self.output_scaler
+            ),
+        )
+        # need a separate model for this so we don't have to
+        # serialize the custom loss functions
+        predict_model = tf.keras.Model(
+            inputs=input_layers,
+            outputs=(T_tendency, q_tendency, surface_precip)
+            + tuple(unpacked_output[2:]),
+        )
+        return train_model, predict_model
+
+    def fit(self, batches: Sequence[xr.Dataset], validation_batch: xr.Dataset) -> None:
+        """Fits a model using data in the batches sequence.
+        """
+        if not self._statistics_are_fit:
+            self.fit_statistics(batches[0])
+        return self._fit_loop(batches, validation_batch)
+
+    def _fit_loop(
+        self, batches: Sequence[xr.Dataset], validation_batch: xr.Dataset
+    ) -> None:
+        if self._train_model is None:
+            self._train_model, self._predict_model = self._build_model()
```

Comment:
Yes, the train and predict model are constructed of the same layers. This was handy in BPTT training because it let us construct a one-timestep model but train over many timesteps, and here it's useful in theory because we can have predictor outputs which are not used in training.

In practice we need to refactor the diagnostic scripts to not crash if there are outputs with no validation data, so for that reason I'll remove the `predict_model` (right now they're identical apart from the loss function being set or not).

Reply:
Ah I was wrong, I wrote a comment reminding me that the train model causes an exception when you try to serialize it because of the loss functions.

I added a docstring to `build_model` saying these share weights (by sharing layers, in this case).